lxd/storage/drivers/drivers_lvm.go

```diff
@@ -0,0 +1,561 @@
+				// Always check if the thin pool exists as we may need to create it later.
+				thinPoolExists, err = d.thinpoolExists(d.config["lvm.vg_name"], d.thinpoolName())
+				if err != nil {
+					return fmt.Errorf("Failed to determine whether thinpool \"%s\" exists in volume group \"%s\": %s", d.config["lvm.vg_name"], d.thinpoolName(), err)
+				}
+
+				// If the single volume is the storage pool's thin pool LV then we still consider
+				// this an empty volume group.
+				if thinPoolExists && lvCount == 1 {
+					empty = true
+				}
+			}
+		} else {
+			empty = true
+		}
+
+		if !empty {
+			return fmt.Errorf("Volume group \"%s\" is not empty", d.config["lvm.vg_name"])
+		}
+
+		// Check the tags on the volume group to check it is not already being used by LXD.
+		if shared.StringInSlice(lvmVgPoolMarker, vgTags) {
+			return fmt.Errorf("Volume group \"%s\" is already used by LXD", d.config["lvm.vg_name"])
+		}
+	} else {
+		// Create physical volume if doesn't exist.
+		if !pvExists {
+			// This is an internal error condition which should never be hit.
+			if pvName == "" {
+				return fmt.Errorf("No name for physical volume detected")
+			}
+
+			_, err := shared.TryRunCommand("pvcreate", pvName)
+			if err != nil {
+				return fmt.Errorf("Failed to create the physical volume for the lvm storage pool: %v", err)
+			}
+			revert.Add(func() { shared.TryRunCommand("pvremove", pvName) })
+		}
+
+		// Create volume group.
+		_, err := shared.TryRunCommand("vgcreate", d.config["lvm.vg_name"], pvName)
+		if err != nil {
+			return fmt.Errorf("Failed to create the volume group for the lvm storage pool: %v", err)
+		}
+		d.logger.Debug("Volume group created", log.Ctx{"pv_name": pvName, "vg_name": d.config["lvm.vg_name"]})
+		revert.Add(func() { shared.TryRunCommand("vgremove", d.config["lvm.vg_name"]) })
+	}
+
+	// Create thin pool if needed.
+	if d.usesThinpool() {
+		if !thinPoolExists {
+			err = d.createThinpool(d.Info().Version, d.config["lvm.vg_name"], d.thinpoolName())
+			if err != nil {
+				return err
+			}
+			d.logger.Debug("Thin pool created", log.Ctx{"vg_name": d.config["lvm.vg_name"], "thinpool_name": d.thinpoolName()})
+
+			revert.Add(func() {
+				d.removeLogicalVolume(d.lvmDevPath(d.config["lvm.vg_name"], "", "", d.thinpoolName()))
+			})
+		}
+	}
+
+	// Mark the volume group with the lvmVgPoolMarker tag to indicate it is now in use by LXD.
+	_, err = shared.TryRunCommand("vgchange", "--addtag", lvmVgPoolMarker, d.config["lvm.vg_name"])
+	if err != nil {
+		return fmt.Errorf("Failed to add marker tag to volume group for the lvm storage pool: %v", err)
+	}
+	d.logger.Debug("LXD marker tag added to volume group", log.Ctx{"vg_name": d.config["lvm.vg_name"]})
+
+	revert.Success()
+	return nil
+}
+
+// Delete removes the storage pool from the storage device.
+func (d *lvm) Delete(op *operations.Operation) error {
+	if d.config["source"] == "" {
+		return fmt.Errorf("No \"source\" property found for the storage pool")
+	}
```

Comment:
I've been dropping such tests from the other drivers. A source is required to create the storage pool and cannot be unset afterwards. If it's not there, the database is corrupted and all bets are off.